### PR TITLE
Add `as_uninit`-like methods to pointer types and unify documentation of `as_ref` methods

### PR DESCRIPTION
**Pull request by TimDiekmann**
This adds a convenient method to retrieve a `&(mut) [MaybeUninit<T>]` from slice pointers (`*const [T]`, `*mut [T]`, `NonNull<[T]>`). See also https://github.com/rust-lang/wg-allocators/issues/66#issuecomment-671789105.

~I'll add a tracking issue as soon as it's reviewed and CI passed.~
Tracking Issue: #75402

r? @RalfJung

### CODE REVIEW

**Comment by RalfJung** [comment]
Is there any particular reason why you added this for slices, but not for individual elements? It seems just as useful for `*const T`/`*mut T`/`NonNull<T>`.

File: library/core/src/ptr/mut_ptr.rs [hunk 1]
```diff
@@ -1112,6 +1112,42 @@ impl<T> *mut [T] {
...
     }
+
+    /// Returns the pointee as a unique, uninitialized slice.
+    ///
+    /// # Safety
+    ///
+    /// Behavior is undefined if any of the following conditions are violated:
+    ///
+    /// * The pointer must be [valid] for reads and writes for `ptr.len() * mem::size_of::<T>()` many bytes,
+    ///   and it must be properly aligned. This means in particular:
```
Comment (RalfJung) [inline library/core/src/ptr/mut_ptr.rs:1123]:
```suggestion
    ///     and it must be properly aligned. This means in particular:
```

**Comment by TimDiekmann** [comment]
That would would Result in `MaybeUninit<[T]>` rather than `[MaybeUninit<T>]`.

**Comment by TimDiekmann** [comment]
Maybe we could also add `as_uninit` for non-slices?

File: library/core/src/ptr/mut_ptr.rs (hunk 1 above)
Comment (TimDiekmann) [reply library/core/src/ptr/mut_ptr.rs:1123]: I'm sorry, I don't understand what you mean. Are you talking about the term "properly aligned" or the indentation of the commentary? According to your edit, you meant the latter, but I don't get it.

**Comment by RalfJung** [comment]
> That would would result in MaybeUninit<[T]> rather than [MaybeUninit<T>].

We probably want a method that converts between those two as well, but I don't know a good way to add such a method. :/

> Maybe we could also add as_uninit for non-slices?

That's what I meant -- something like `as_uninit`/`as_uninit_ref` or so.

File: library/core/src/ptr/mut_ptr.rs (hunk 1 above)
Comment (RalfJung) [reply library/core/src/ptr/mut_ptr.rs:1123]: Ah lol "alignment" is ambiguous. ;)

I meant the indention. But I just realized you probably deliberately made these next two bullets further indented as sub-bullets. Let's hope this renders correctly, which I find really hard to predict with markdown.

File: library/core/src/ptr/const_ptr.rs [hunk 2]
```diff
@@ -906,6 +906,41 @@ impl<T> *const [T] {
...
+    ///   See the safety documentation of [`pointer::offset`].
+    ///
+    /// See also [`slice::from_raw_parts`].
+    ///
+    /// [valid]: ../index.html#safety
+    /// [`NonNull::dangling()`]: NonNull::dangling
+    /// [`pointer::offset`]: ../primitive.pointer.html#method.offset
+    #[inline]
+    #[unstable(feature = "uninit_ptr_slice", issue = "none")]
+    pub unsafe fn as_uninit_slice<'a>(self) -> &'a [MaybeUninit<T>] {
```
Comment (RalfJung) [inline library/core/src/ptr/const_ptr.rs:940]: Please note in the tracking issue that the naming is different from the usual raw ptr methods which indicate mutability -- like `from_raw_parts_mut` or `as_mut`. I don't have a strong opinion either way, but the libs team might. ;)

Comment (TimDiekmann) [reply library/core/src/ptr/const_ptr.rs:940]: I could rename the feature to `slice_ptr_uninit` which follows the naming scheme of #74265.

Comment (RalfJung) [reply library/core/src/ptr/const_ptr.rs:940]: I don't think anyone cares about feature names, but sure.^^
I mean the names of the methods.

Comment (TimDiekmann) [reply library/core/src/ptr/const_ptr.rs:940]: Oh sure :smile: 

Well, currently the term "uninit" is always prepended to the type: https://doc.rust-lang.org/nightly/core/index.html?search=uninit

Comment (RalfJung) [reply library/core/src/ptr/const_ptr.rs:940]: I was talking about the fact that the const and mut methods have the same name, which they usually do not.

**Comment by TimDiekmann** [comment]
> We probably want a method that converts between those two as well, but I don't know a good way to add such a method. :/

Me neither, we probably want a method converting in both direction with a decent name...

> That's what I meant -- something like `as_uninit`/`as_uninit_ref` or so.

I'd name them `as_uninit_ref` to make them analogue to `as_ref`. Do we want another pull request for this or should I just add this here, gated behind `as_uninit_ref`? Should I add both to the same tracking issue?

**Comment by RalfJung** [comment]
I'd make them all one PR and one feature.

File: library/core/src/ptr/mut_ptr.rs (hunk 1 above)
Comment (TimDiekmann) [reply library/core/src/ptr/mut_ptr.rs:1123]: The comment is largely copied from `slice::from_raw_parts(_mut)`, so it should work :)

**Comment by TimDiekmann** [comment]
Currently we have `NonNull::as_ref` and `NonNull::as_mut` taking `&(mut) self` (both stabilized - sadly). Shall I also alter `as_uninit_slice` to take `&self` and returning `&[MaybeUninit]` and add `as_uninit_slice_mut`? Since `NonNull` implements `Copy` I don't like passing it as a reference, as you will almost never have a reference to `NonNull`, especially not a `&mut NonNull`.

This btw. also applies to `as_mut_ptr` which converts a slice to a non-slice pointer. On nightly there is `MaybeUninit::first_ptr(_mut)` which maybe could also be used on slice pointers instead. With this naming we could keep a single naming scheme.

File: library/core/src/ptr/const_ptr.rs (hunk 2 above)
Comment (TimDiekmann) [reply library/core/src/ptr/const_ptr.rs:940]: Oh :D
I'll change this.

**Comment by RalfJung** [comment]
> Shall I also alter as_uninit_slice to take &self and returning &[MaybeUninit] and add as_uninit_slice_mut?

I can do technical reviews but this is a question for @rust-lang/libs that I cannot help with, I am afraid.

**Comment by TimDiekmann** [comment]
I'll first implement the methods as they were already implemented with the pointer types. Since some methods are already stable, I think that a unified schema is the most important thing.

**Comment by TimDiekmann** [comment]
I just noticed, that `pointer::as_ref` returns `Option<&T>`.

**Comment by TimDiekmann** [comment]
> Is there any particular reason why you added this for slices, but not for general pointers? It seems just as useful for `*const T`/`*mut T`/`NonNull<T>`.

One thing to note: `MaybeUninit<[T]>` isn't even possible right now, as `MaybeUninit<T>` requires `T` to be sized.
Noticed that when I implemented `as_uninit_ref/mut` :D

**Comment by RalfJung** [comment]
Ah good point, and we are unlikely to be able to relax that sized bound.

File: library/core/src/ptr/const_ptr.rs [hunk 3]
```diff
@@ -101,6 +100,35 @@ impl<T: ?Sized> *const T {
         if self.is_null() { None } else { unsafe { Some(&*self) } }
     }
 
+    /// Returns `None` if the pointer is null, or else returns a reference to
+    /// the value wrapped in `Some`.
+    ///
+    /// While this method is useful for null-safety, it is important to note
+    /// that this is still an unsafe operation because the returned value could
+    /// be pointing to invalid memory.
```
Comment (RalfJung) [inline library/core/src/ptr/const_ptr.rs:108]: "Invalid memory" in the sense of "containing invalid data" is not a concern.

But the pointer could be a non-null dangling pointer. That would be a problem.

File: library/core/src/ptr/const_ptr.rs [hunk 4]
```diff
@@ -101,6 +100,35 @@ impl<T: ?Sized> *const T {
...
+    ///
+    /// # Safety
+    ///
+    /// When calling this method, you have to ensure that *either* the pointer is NULL *or*
+    /// it is properly aligned
```
Comment (RalfJung) [inline library/core/src/ptr/const_ptr.rs:113]: It must also be dereferencable.

File: library/core/src/ptr/mut_ptr.rs [hunk 5]
```diff
@@ -100,6 +100,36 @@ impl<T: ?Sized> *mut T {
...
+    ///
+    /// # Safety
+    ///
+    /// While this method and its mutable counterpart are useful for
+    /// null-safety, it is important to note that this is still an unsafe
+    /// operation because the returned value could be pointing to invalid
+    /// memory.
+    ///
+    /// When calling this method, you have to ensure that *either* the pointer is NULL *or*
+    /// it is properly aligned
```
Comment (RalfJung) [inline library/core/src/ptr/mut_ptr.rs:114]: Likewise, it must also be dereferencable.

File: library/core/src/ptr/mut_ptr.rs [hunk 6]
```diff
@@ -100,6 +100,36 @@ impl<T: ?Sized> *mut T {
         if self.is_null() { None } else { unsafe { Some(&*self) } }
     }
 
+    /// Returns `None` if the pointer is null, or else returns a reference to
+    /// the value wrapped in `Some`.
+    ///
+    /// # Safety
+    ///
+    /// While this method and its mutable counterpart are useful for
```
Comment (RalfJung) [inline library/core/src/ptr/mut_ptr.rs:108]: Above you removed the "and its counterpart", I think it would make sense to also remove it here (and in `as_ref`). Alternatively, link to what that counterpart is.

File: library/core/src/ptr/mut_ptr.rs [hunk 7]
```diff
@@ -286,6 +316,37 @@ impl<T: ?Sized> *mut T {
...
+    /// As with [`as_uninit_ref`], this is unsafe because it cannot verify the validity
+    /// of the returned pointer, nor can it ensure that the lifetime `'a`
+    /// returned is indeed a valid lifetime for the contained data.
+    ///
+    /// [`as_uninit_ref`]: #method.as_uninit_ref
+    ///
+    /// # Safety
+    ///
+    /// When calling this method, you have to ensure that *either* the pointer is NULL *or*
+    /// it is properly aligned
```
Comment (RalfJung) [inline library/core/src/ptr/mut_ptr.rs:330]: Likewise it must be dereferencable.

File: library/core/src/ptr/non_null.rs [hunk 8]
```diff
@@ -76,6 +76,54 @@ impl<T: Sized> NonNull<T> {
...
+
+    /// Dereferences the content.
+    ///
+    /// The resulting lifetime is bound to self so this behaves "as if"
+    /// it were actually an instance of T that is getting borrowed. If a longer
+    /// (unbound) lifetime is needed, use `&*my_ptr.as_ptr()`.
+    ///
+    /// # Safety
+    ///
+    /// When calling this method, you have to ensure that `self` is properly aligned.
```
Comment (RalfJung) [inline library/core/src/ptr/non_null.rs:88]: and it must be dereferencable

File: library/core/src/ptr/non_null.rs [hunk 9]
```diff
@@ -76,6 +76,54 @@ impl<T: Sized> NonNull<T> {
...
+
+    /// Mutably dereferences the content.
+    ///
+    /// The resulting lifetime is bound to self so this behaves "as if"
+    /// it were actually an instance of T that is getting borrowed. If a longer
+    /// (unbound) lifetime is needed, use `&mut *my_ptr.as_ptr()`.
+    ///
+    /// # Safety
+    ///
+    /// When calling this method, you have to ensure that `self` is properly aligned.
```
Comment (RalfJung) [inline library/core/src/ptr/non_null.rs:112]: and it must be dereferencable

File: library/core/src/ptr/mut_ptr.rs [hunk 10]
```diff
@@ -105,20 +106,21 @@ impl<T: ?Sized> *mut T {
     ///
     /// # Safety
     ///
-    /// While this method and its mutable counterpart are useful for
+    /// While this method and its [mutable counterpart] are useful for
```
Comment (RalfJung) [inline library/core/src/ptr/mut_ptr.rs:109]: For symmetry I would then expect the same wording on that mutable counterpart -- something like "this method and its shared counterpart". Doesn't seem like that's what we got though? (It's a bit hard to tell in the GH diff.)

**Comment by TimDiekmann** [comment]
Those intra-doc links drive me crazy :D
CI should pass now.

File: library/core/src/ptr/const_ptr.rs [hunk 11]
```diff
@@ -906,6 +933,50 @@ impl<T> *const [T] {
...
+
+    /// Returns `None` if the pointer is null, or else returns a shared reference to
+    /// the value wrapped in `Some`.
+    ///
+    /// # Safety
+    ///
+    /// Behavior is undefined if any of the following conditions are violated:
+    ///
+    /// * The pointer must be [valid] for reads for `ptr.len() * mem::size_of::<T>()` many bytes,
+    ///   and it must be properly aligned. This means in particular:
```
Comment (RalfJung) [inline library/core/src/ptr/const_ptr.rs:945]: The requirement for `as_uninit_ref` and `as_uninit_slice` is basically the same -- aligned and dereferencable -- but it is stated very differently. I am not sure what is better, but this is a curious inconsistency.

Comment (TimDiekmann) [reply library/core/src/ptr/const_ptr.rs:945]: I personally like the wording of `as_uninit_slice` better. It is easier to skim a list of bullet points than to read a coherent text.

Comment (RalfJung) [reply library/core/src/ptr/const_ptr.rs:945]: Agreed. That's also how [`ptr:.write`](https://doc.rust-lang.org/nightly/std/ptr/fn.write.html) and similar methods look.

Comment (TimDiekmann) [reply library/core/src/ptr/const_ptr.rs:945]: Do you want me to change the docs for `as_ref` and `as_mut`, too? Then they are unified.

Comment (RalfJung) [reply library/core/src/ptr/const_ptr.rs:945]: That would be great if you could do that. :)

Comment (TimDiekmann) [reply library/core/src/ptr/const_ptr.rs:981]: I think I have everything added. Please review the safety clauses carefully, let's don't risk I forgot something :)

File: library/core/src/ptr/const_ptr.rs [hunk 12]
```diff
@@ -38,32 +38,30 @@ impl<T: ?Sized> *const T {
...
     /// all of the following is true:
-    /// - it is properly aligned
-    /// - it must point to an initialized instance of T; in particular, the pointer must be
-    ///   "dereferenceable" in the sense defined [here].
+    ///
+    /// * The pointer must be properly aligned,
+    ///
+    /// * it must be "dereferencable" in the sense defined in [the module documentation], and
+    ///
+    /// * the lifetime `'a` returned is arbitrarily chosen and does
```
Comment (RalfJung) [inline library/core/src/ptr/const_ptr.rs:53]: The wording is odd. You must ensure that the lifetime is arbitrarily chosen? I'd start with something like "*You* must enforce Rust's aliasing rules, since the returned lifetime `'a` is arbitrarily chosen. In particular, ..."

File: library/core/src/ptr/const_ptr.rs [hunk 13]
```diff
@@ -38,32 +38,30 @@ impl<T: ?Sized> *const T {
...
+    ///   not necessarily reflect the actual lifetime of the data. *You* must enforce
+    ///   Rust's aliasing rules. In particular, for the duration of this lifetime,
+    ///   the memory the pointer points to must not get mutated (except inside `UnsafeCell`).
+    ///
+    /// * the pointer must point to an initialized instance of `T`.
```
Comment (RalfJung) [inline library/core/src/ptr/const_ptr.rs:58]: I'd move this up above the lifetime.

File: library/core/src/ptr/const_ptr.rs [hunk 14]
```diff
@@ -101,6 +99,70 @@ impl<T: ?Sized> *const T {
...
+    /// `as_uninit_ref_unchecked` that returns the `&MaybeUninit<T>` instead of
+    /// `Option<&MaybeUninit<T>>`, know that you can dereference the pointer directly.
+    ///
+    /// ```
+    /// use std::mem::MaybeUninit;
+    ///
+    /// let ptr: *const u8 = &10u8 as *const u8;
+    ///
+    /// unsafe {
+    ///     let val_back = &*(ptr as *const MaybeUninit<u8>);
```
Comment (RalfJung) [inline library/core/src/ptr/const_ptr.rs:151]: I am always hesitant to recommend pointer casts. They are a very dangerous operation as the type checker checks very little about them.

There's no unwrap_unchecked, is there?

TBH I never understood why NULL was treated separately from the other erorr conditions in these methods, but I expect it is too confusing to make the uninit versions return a plain reference without `Option`?

Comment (TimDiekmann) [reply library/core/src/ptr/const_ptr.rs:151]: I just copied the previous example but I don't like the example either. Actually the method is unsafe anyway, so dereferencing a raw pointer in this context seems ok'ish.

No, sadly, there is no `unwrap_unchecked` but there should be IMO. If we add it, this example should definitely be changed.

Yes, that was the reason about it. I'd expect `as_uninit_ref` to behave exactly as `as_ref`, besides wrapped in an `MaybeUninit`. Also it is UB to have `&MaybeUninit<T>` pointing to NULL. For a pointer guaranteed to be non-NULL we have `NonNull`, which should be used whenever possible IMO.

Comment (RalfJung) [reply library/core/src/ptr/const_ptr.rs:151]: The example you copied doesn't need a pointer cast though.

Comment (TimDiekmann) [reply library/core/src/ptr/const_ptr.rs:151]: Oh, I'm sorry. Maybe we should just leave that out then?

Comment (RalfJung) [reply library/core/src/ptr/const_ptr.rs:151]: It's still a fair point though, there should be an ergonomic way to do that, and IMO "ergonomic" implies "no `as`".

I am fine with punting on that for now and leaving a note in the tracking issue.

File: library/core/src/ptr/const_ptr.rs [hunk 15]
```diff
@@ -38,32 +38,30 @@ impl<T: ?Sized> *const T {
...
-    ///
     /// When calling this method, you have to ensure that *either* the pointer is NULL *or*
     /// all of the following is true:
-    /// - it is properly aligned
-    /// - it must point to an initialized instance of T; in particular, the pointer must be
-    ///   "dereferenceable" in the sense defined [here].
+    ///
+    /// * The pointer must be properly aligned,
+    ///
+    /// * it must be "dereferencable" in the sense defined in [the module documentation], and
```
Comment (RalfJung) [inline library/core/src/ptr/const_ptr.rs:51]: You have a ", and" here but not on the first item. Given that the list starts with "all of the following must be true", I don't think the ", and" is necessary, but if you want to use it please do so consistently.

File: library/core/src/ptr/const_ptr.rs [hunk 16]
```diff
@@ -101,6 +99,53 @@ impl<T: ?Sized> *const T {
         if self.is_null() { None } else { unsafe { Some(&*self) } }
     }
 
+    /// Returns `None` if the pointer is null, or else returns a shared reference to
+    /// the value wrapped in `Some`.
```
Comment (RalfJung) [inline library/core/src/ptr/const_ptr.rs:103]: This should probably mention "uninitialized" somewhere. It could point out explicitly that the key difference to `as_ref`. is that the pointed-to data does *not* have to be initialized.

File: library/core/src/ptr/const_ptr.rs [hunk 17]
```diff
@@ -906,6 +951,52 @@ impl<T> *const [T] {
...
+    ///       Slices can never span across multiple allocated objects, and
+    ///
+    ///     * The pointer must be aligned even for zero-length slices. One
+    ///       reason for this is that enum layout optimizations may rely on references
+    ///       (including slices of any length) being aligned and non-null to distinguish
+    ///       them from other data. You can obtain a pointer that is usable as `data`
+    ///       for zero-length slices using [`NonNull::dangling()`].
+    ///
+    /// * The total size `ptr.len() * mem::size_of::<T>()` of the slice must be no larger than `isize::MAX`.
+    ///   See the safety documentation of [`pointer::offset`],
```
Comment (RalfJung) [inline library/core/src/ptr/const_ptr.rs:976]: Ending this in a "," even though this is a two-sentence bullet is somewhat odd.

**Comment by RalfJung** [comment]
In terms of text this is all fine (assuming it's the same text for correspond methods, which I did not check in detail everywhere). I just have some more typographic nits.

File: library/core/src/ptr/const_ptr.rs (hunk 16 above)
Comment (TimDiekmann) [reply library/core/src/ptr/const_ptr.rs:103]: May I add something like "In comparison to [`as_ref`](link to as_ref), this does not require, that the value has to be initialized."?

File: library/core/src/ptr/const_ptr.rs (hunk 15 above)
Comment (TimDiekmann) [reply library/core/src/ptr/const_ptr.rs:51]: The " and" was a copy-paste misstake, thanks!

File: library/core/src/ptr/const_ptr.rs (hunk 16 above)
Comment (RalfJung) [reply library/core/src/ptr/const_ptr.rs:103]: I'd probably say "in contrast to", and the last "," doesn't belong there I think, but yes. :)

**Comment by TimDiekmann** [comment]
Thank you!
I'll double-check every text and it's method again and will push it later this day.

File: library/core/src/ptr/const_ptr.rs (hunk 15 above)
Comment (RalfJung) [reply library/core/src/ptr/const_ptr.rs:51]: Basically these should be either self-contained sentences (starting with a capital letter and ending in a `.`), or forma  long sentence together. But the last gets kind of awkward given that the last item is multiple sentences long.

File: library/core/src/ptr/const_ptr.rs (hunk 16 above)
Comment (TimDiekmann) [reply library/core/src/ptr/const_ptr.rs:103]: Alright. Well, the last "," is German grammar I think 😄

File: library/core/src/ptr/const_ptr.rs (hunk 15 above)
Comment (TimDiekmann) [reply library/core/src/ptr/const_ptr.rs:51]: Is there a guideline for writing safety-sections somewhere? It seems that there are several approaches, even in `std`. A few examples:
- [`str::get_unchecked`](https://doc.rust-lang.org/nightly/std/primitive.str.html#method.get_unchecked) uses ";" and "."
- [`slice::from_raw_parts`](https://doc.rust-lang.org/nightly/core/slice/fn.from_raw_parts.html) or [`str::slice_unchecked`](https://doc.rust-lang.org/nightly/std/primitive.str.html#method.slice_unchecked) uses "."
- [`GlobalAlloc::realloc`](https://doc.rust-lang.org/nightly/core/alloc/trait.GlobalAlloc.html#method.realloc) (and other methods from `core::alloc`) uses a mix of ", and", ",", and "."

Comment (RalfJung) [reply library/core/src/ptr/const_ptr.rs:51]: I don't think there is... I'd have used `.` for all of those.

File: library/core/src/ptr/const_ptr.rs (hunk 16 above)
Comment (TimDiekmann) [reply library/core/src/ptr/const_ptr.rs]: I have also added a link from `as_ref` to `as_uninit_ref` to quickly jump back.

File: library/core/src/ptr/const_ptr.rs [hunk 18]
```diff
@@ -966,16 +975,16 @@ impl<T> *const [T] {
     ///     * The entire memory range of this slice must be contained within a single allocated object!
     ///       Slices can never span across multiple allocated objects, and
     ///
-    ///     * The pointer must be aligned even for zero-length slices. One
+    ///     * the pointer must be aligned even for zero-length slices. One
```
Comment (RalfJung) [inline library/core/src/ptr/const_ptr.rs]: "Slices can never span across multiple allocated objects, and the pointer must be aligned even for zero-length slices." -- that's an odd sentence, so I think this is worth changing to proper sentences as well.

(I hope I am not nitpicking too much here, if you want we can also land this and do a cleanup/editing PR afterwards.)

Comment (TimDiekmann) [reply library/core/src/ptr/const_ptr.rs]: That review is already outdated ;)

Comment (RalfJung) [reply library/core/src/ptr/const_ptr.rs]: Ah, I got no further email from GH but I guess you force-pushed?

File: library/core/src/ptr/const_ptr.rs [hunk 19]
```diff
@@ -906,6 +957,55 @@ impl<T> *const [T] {
...
+    /// When calling this method, you have to ensure that *either* the pointer is NULL *or*
+    /// all of the following is true:
+    ///
+    /// * The pointer must be [valid] for reads for `ptr.len() * mem::size_of::<T>()` many bytes,
+    ///   and it must be properly aligned. This means in particular:
+    ///
+    ///     * The entire memory range of this slice must be contained within a single allocated object!
+    ///       Slices can never span across multiple allocated objects, and
+    ///
+    ///     * the pointer must be aligned even for zero-length slices. One
```
Comment (RalfJung) [inline library/core/src/ptr/const_ptr.rs:978]: Still looks like such an odd sentence here though?

File: library/core/src/ptr/const_ptr.rs (hunk 18 above)
Comment (TimDiekmann) [reply library/core/src/ptr/const_ptr.rs]: No, I think your diff was outdated. The changes where pushed with 82b8d75 :)
I try not to force push whenever possible.

File: library/core/src/ptr/const_ptr.rs (hunk 19 above)
Comment (TimDiekmann) [reply library/core/src/ptr/const_ptr.rs:978]: For me it felt natural to continue the sentence there, since both complement the previous sentence, but I think that's a matter of taste. If you want, I can change that.

Comment (RalfJung) [reply library/core/src/ptr/const_ptr.rs:978]: The previous sentence here is "The entire memory range of this slice must be contained within a single allocated object!". I don't see how the second bullet point complements that point.

The two bullets are two separate independent points, and yet somehow the second half of the first bullet and the beginning of the second bullet are glued into one sentence. I don't see how that makes any sense.

Comment (RalfJung) [reply library/core/src/ptr/const_ptr.rs:978]: Maybe you mean that both complement "The pointer must be valid for reads", but that is not what the grammatical structure indicates here. The sentence structure indicates a connection between "must be contained within a single allocated object" on the one hand, and ("never span across multiple allocated objects" + "must be aligned even for zero-length slices") on the other hand, which to me just makes no sense.

Comment (TimDiekmann) [reply library/core/src/ptr/const_ptr.rs:978]: Well, I think you are right. Reinvestigating [`slice::from_raw_parts`](https://doc.rust-lang.org/nightly/core/slice/fn.from_raw_parts.html) shows, that its documentation also uses two separate sentences.

Comment (TimDiekmann) [reply library/core/src/ptr/const_ptr.rs:978]: Do you think it is enough if I change it as follows, or should it be wound up a bit more?
```suggestion
    ///       Slices can never span across multiple allocated objects.
    ///
    ///     * The pointer must be aligned even for zero-length slices. One
```

Comment (RalfJung) [reply library/core/src/ptr/const_ptr.rs:978]: That diff seems fine to me. What other changes were you thinking of?

Comment (TimDiekmann) [reply library/core/src/ptr/const_ptr.rs:978]: I do not have any concrete ones in mind at the moment, maybe you can think of something else?

Comment (RalfJung) [reply library/core/src/ptr/const_ptr.rs:978]: Not really, that's why I am asking. ;)
Let's just go with that diff.

File: library/core/src/ptr/non_null.rs [hunk 20]
```diff
@@ -278,6 +352,114 @@ impl<T> NonNull<[T]> {
...
+    ///
+    /// ```rust
+    /// #![feature(allocator_api, ptr_as_uninit)]
+    ///
+    /// use std::alloc::{AllocRef, Layout, Global};
+    /// use std::mem::MaybeUninit;
+    /// use std::ptr::NonNull;
+    ///
+    /// let memory: NonNull<[u8]> = Global.alloc(Layout::new::<[u8; 32]>())?;
+    /// // This is safe as `memory` is a valid pointer to `[u8]`
```
Comment (RalfJung) [inline library/core/src/ptr/non_null.rs:451]: It's not initialized though, so it's not really valid?

Comment (TimDiekmann) [reply library/core/src/ptr/non_null.rs:451]: From `core::ptr#Safety`:

> For a pointer to be valid, it is necessary, but not always sufficient, that the pointer be dereferenceable: the memory range of the given size starting at the pointer must all be within the bounds of a single allocated object.

The safety constraint from `AllocRef` requires this:
> Memory blocks returned from an allocator must point to valid memory.

We have to elaborate on this in the allocator-wg, but this is another topic. Basically it is required for all allocators, that a valid pointer is returned. For this, there are currently two issues open: https://github.com/rust-lang/wg-allocators/issues/65 https://github.com/rust-lang/wg-allocators/issues/46. The former wants to refine the quoted phrase from `AllocRef`.

To sum up: This is not initialized, but valid. As long as we don't return `&[T]` this should be fine.

, that it's dereferencable. An allocator returns a well-aligned pointer to an allocated object. The pointer does not necessarily need to point to initialized memory for this method to be called. I think this is the whole point of `MaybeUninit`.

Comment (RalfJung) [reply library/core/src/ptr/non_null.rs:451]: That is *valid for a given access kind and size*. But what your text here says sounds more like "valid for a given type", which -- for me at least -- is something very different.

Comment (TimDiekmann) [reply library/core/src/ptr/non_null.rs:451]: I don't think I quite follow. Which validity invariant am I missing here?

The whole reason to use an allocator (and let's assume that `Global` is implemented correctly here) is to get back a memory region that we can write to. Such a memory area may contain anything, uninitialized, but has a fixed size of at least `NonNull::len()`. For me this is just like returning a slice of `MaybeUninit<u8>` with a size of `len()`. Since `alloc` returns `NonNull<[u8]>`, it is guaranteed that no padding exists between `ptr[0]` and `ptr[1]` as [`u8` has no padding bits and its size is exactly one byte](https://rust-lang.github.io/unsafe-code-guidelines/layout/scalars.html#fixed-width-integer-types). So why is `&[MaybeUninit<u8>]` not a valid representation?

It may well be that I overlook or misunderstand something, but at the moment I cannot see the reason.

File: library/core/src/ptr/non_null.rs [hunk 21]
```diff
@@ -278,6 +352,114 @@ impl<T> NonNull<[T]> {
...
+    /// For the shared counterpart see [`as_uninit_slice`].
+    ///
+    /// [`as_mut`]: #method.as_mut
+    /// [`as_uninit_slice`]: #method.as_uninit_slice
+    ///
+    /// # Safety
+    ///
+    /// When calling this method, you have to ensure that all of the following is true:
+    ///
+    /// * The pointer must be [valid] for reads for `ptr.len() * mem::size_of::<T>()` many bytes,
```
Comment (RalfJung) [inline library/core/src/ptr/non_null.rs:413]: "reads" and writes?

File: library/core/src/ptr/mut_ptr.rs [hunk 22]
```diff
@@ -1112,6 +1211,110 @@ impl<T> *mut [T] {
...
+    ///
+    /// [`as_mut`]: #method.as_mut
+    /// [`as_uninit_slice`]: #method.as_uninit_slice-1
+    ///
+    /// # Safety
+    ///
+    /// When calling this method, you have to ensure that *either* the pointer is NULL *or*
+    /// all of the following is true:
+    ///
+    /// * The pointer must be [valid] for reads for `ptr.len() * mem::size_of::<T>()` many bytes,
```
Comment (RalfJung) [inline library/core/src/ptr/mut_ptr.rs:1281]: and writes

File: library/core/src/ptr/non_null.rs (hunk 20 above)
Comment (RalfJung) [reply library/core/src/ptr/non_null.rs:451]: > Which validity invariant am I missing here?

`u8` must be initialized. So allocators do not return valid `u8`.

Your safety docs get it right, it says

> The pointer must be valid for reads (and writes) for `ptr.len() * mem::size_of::<T>()` many bytes

So the code's safety comment should match those docs. But they do not.
The correct thing to say would be something like "`memory` is valid for read and writes for `memory.len()` many bytes".

File: library/core/src/ptr/non_null.rs (hunk 21 above)
Comment (TimDiekmann) [reply library/core/src/ptr/non_null.rs:413]: Oh, yes 😄

File: library/core/src/ptr/non_null.rs (hunk 20 above)
Comment (TimDiekmann) [reply library/core/src/ptr/non_null.rs:451]: Oh, I'm so glad. I was a little shocked :D

Sure, the comment is very poor worded and "valid" isn't the right term here. I think your suggestion makes it more clear:

```suggestion
    /// // This is safe as `memory` is a valid for reads and writes for `memory.len()` many bytes.
    /// // Note, that calling `memory.as_mut()` is not allowed here as the content may be uninitialized.
```

Comment (RalfJung) [reply library/core/src/ptr/non_null.rs:451]: "is a valid" -> "is valid"

Also I don't think there should be a comma after "Note".

**Comment by RalfJung** [comment]
All right, I think this is ready. :) Could you squash the commits?

**Comment by TimDiekmann** [comment]
Done (in case GH does not send mails on force-pushes)

**Comment by RalfJung** [comment]
> Done (in case GH does not send mails on force-pushes).

(it doesn't, except sometimes when it goes... their support told me it only sends them when the parent of the new HEAD was already in the branch, aka when you just amended the last commit.)

Great, and thanks for enduring by never-ending stream of editorial comments. :)
@bors r+ rollup

**Comment by TimDiekmann** [comment]
> Great, and thanks for enduring by never-ending stream of editorial comments. :)

I think the effort was worth it :)
